format/spec.md

```diff
@@ -193,6 +193,17 @@ Notes:
 
 For details on how to serialize a schema to JSON, see Appendix C.
 
+#### Default value
+
+Default value can be assigned to a column when the column is added to an Iceberg table as part of the schema evolution. They are tracked at the level of a nested field inside a struct, thus it can be used for both top-level columns and nested columns. Iceberg tracks two default values internally: `initial-default` and `write-default`. The `initial-default` is used to read rows belonging to files that lack the column (i.e. the files were written before the column is added); the `write-default` value will be used for the automatically populating the column if user later inserts new rows without specifying the column.
```

Comment:
Beginning when "assigned ... when ..." is not giving context about what field defaults are. This should begin with a high-level explanation and then go into detail. Also, this is the Iceberg spec so it doesn't make sense to refer to what "Iceberg tracks". Instead, this should state what is required.

> Default values can be tracked for struct fields (both nested structs and the top-level schema's struct). There are two defaults for a field:
> * `initial-default` is a value that must be projected when reading a data file that was written before the field was added to the schema
> * `write-default` is a value that must be written for all rows when the field is missing from input data while writing a new data file
>
> Note that all schema fields are required when writing data into an Iceberg table. Omitting a known field from a data file is not allowed. The write default for a field should be written when a field is not supplied to a write. If the write default is not set, the writer must fail.

Reply:
A couple of suggestions for the names: `file-to-record default` and `record-to-file default`, or `read-time default` and `write-time default`. I feel `file-to-record default` and `record-to-file default` are expressive, symmetric, and accurately capture the function of each default. We might explain that the former cannot be changed, while the latter can be changed throughout the lifecycle of a table. The former cannot be changed because it is used to read existing files while the latter can be changed because it is used to write new files.

We might also start by explaining the semantics of default values from a _contract_ point of view:

> A default value associated with a field is used to:
> (1) populate the field's value for all records that were written before the field is introduced
> (2) populate the field's value for any records that will be written after the field is introduced, when such records do not supply that field's value.
> All fields introduced at table creation time (i.e., not part of schema evolution) only leverage the second use case. Fields introduced as part of a schema evolution can leverage both use cases.
> 
> Changes of default values apply to future records only (i.e., records written after the default value has changed). To prevent retroactive change of record values that are already written, each field is associated with two default values ... (then we can continue with the section above).

Reply:
@wmoustafa, I think that these should _not_ be symmetric because they have very different behaviors. The initial default can only be set when the column is created. The write default can change any time. Also, the initial default is explicitly not a read-time default. While it is applied at read time, I think that `file-to-record` or `read-time` imply that the default can be "used" by not writing columns into data files, which is explicitly disallowed.

I'd keep the `initial-default` and `write-default` names for clarity.

Reply:
@rdblue I am fine with that. Could you comment on the intro (the second paragraph in my comment)?

Reply:
@rdblue Updated the spec according to the above discussions, could you take a look again?